### PR TITLE
Normalize keys in rechunk

### PR DESCRIPTION
**Pull request by mrocklin**
- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

### CODE REVIEW

**Comment by mrocklin** [comment]
Merging tomorrow if there are no comments

File: dask/array/rechunk.py [hunk 1]
```diff
@@ -214,7 +213,16 @@ def rechunk(x, chunks, threshold=None,
...
     if isinstance(chunks, dict):
-        chunks = dict(chunks)
+        new = {}
+        for k, v in chunks.items():
+            if not isinstance(k, int) or k >= x.ndim or k <= -x.ndim:
+                msg = ("Invalid dimension %s. Should be an integer from 0 to %d"
+                       % (k, x.ndim))
+                raise ValueError(msg)
+            if k < 0:
+                k += x.ndim
```
Comment (TomAugspurger) [inline dask/array/rechunk.py]: Thoughts on raising here if that would clobber a key in `new`? e.g. the user wrote `x.rechunk({2: 1, -1: 1})` for a 2-D array.

Comment (mrocklin) [reply dask/array/rechunk.py]: Meh.  I say we wait until this actually happens.

File: dask/array/rechunk.py [hunk 2]
```diff
@@ -214,7 +213,16 @@ def rechunk(x, chunks, threshold=None,
     >>> y = x.rechunk({0: -1, 1: 'auto'}, block_size_limit=1e8)
     """
     if isinstance(chunks, dict):
-        chunks = dict(chunks)
+        new = {}
+        for k, v in chunks.items():
+            if not isinstance(k, int) or k >= x.ndim or k <= -x.ndim:
```
Comment (shoyer) [inline dask/array/rechunk.py]: `k=-x.ndim` is valid, e.g., `{-2: 1}` for a 2D array. I usually use `not -ndim <= axis < ndim` for this check.

Comment (shoyer) [reply dask/array/rechunk.py]: e.g., see https://github.com/pydata/xarray/blob/1688a59803786a9d88eeb43aa4c935f7052d6a80/xarray/core/nputils.py#L9

**Comment by mrocklin** [comment]
Inspired by XArray I've standardized these checks into a `validate_axis` function

**Pull request merged**